Code/GraphMol/Fingerprints/MHFP.h

```diff
@@ -0,0 +1,302 @@
+       const bool& kekulize = true,
+       const unsigned char& min_radius = 1);
+
+
+  /*!
+    \brief Creates a binary fingerprint based on circular sub-SMILES.
+   
+    Creates a binary fingerprint similar to ECFP. However, instead of using
+    a Morgan-style hashing, circular n-grams (sub-SMILES) are created, hashed
+    directly and folded.
+   
+    \param radius the maximum radius of the substructure that is generated at
+          each atom. Default: <tt>3</tt>.
+    \param rings whether the rings (SSSR) are extrected from the molecule and
+           added to the shingling. Given the molecule 
+           <tt>"C1CCCCCC1C(=O)C"</tt>, "<tt>C1CCCCCC1"</tt> would be added
+           to the shingling. Default: <tt>true</tt>.
+    \param isomeric whether the SMILES added to the shingling are isomeric.
+            Default: <tt>false</tt>.
+    \param kekulize whether the SMILES added to the shingling are kekulized.
+            Default: <tt>true</tt>.
+    \param min_radius the minimum radius that is used to extract n-grams.
+            Default: <tt>1</tt>.
+    \param length the length into which the fingerprint is folded.
+            Default: <tt>2048</tt>.
+
+    \returns the SECFP fingerprint.
+   */
+  ExplicitBitVect
+  EncodeSECFP(ROMol& mol,
+       const unsigned char& radius = 3,
+       const bool& rings = true,
+       const bool& isomeric = false,
+       const bool& kekulize = true,
+       const unsigned char& min_radius = 1,
+       const size_t& length = 2048);
+
+  //! \overload
+  std::vector<ExplicitBitVect>
+  EncodeSECFP(std::vector<ROMol>& mols,
+       const unsigned char& radius = 3,
+       const bool& rings = true,
+       const bool& isomeric = false,
+       const bool& kekulize = true,
+       const unsigned char& min_radius = 1,
+       const size_t& length = 2048);
+
+  //! \overload
+  ExplicitBitVect
+  EncodeSECFP(std::string& smiles,
+       const unsigned char& radius = 3,
+       const bool& rings = true,
+       const bool& isomeric = false,
+       const bool& kekulize = true,
+       const unsigned char& min_radius = 1,
+       const size_t& length = 2048);
+
+  //! \overload
+  std::vector<ExplicitBitVect>
+  EncodeSECFP(std::vector<std::string>& smiles,
+       const unsigned char& radius = 3,
+       const bool& rings = true,
+       const bool& isomeric = false,
+       const bool& kekulize = true,
+       const unsigned char& min_radius = 1,
+       const size_t& length = 2048);
+
+  /*!
+    \brief Calculates the Jaccard / Tanimoto distance between two MHFP fingerprints.
+   
+    \param a an MHFP fingerprint vector.
+    \param b an MHFP fingerprint vector.
+
+    \returns the Jaccard / Tanimoto distance between the two fingerprints.
+   */
+  static float
+  Distance(const std::vector<uint32_t>& a, 
+       const std::vector<uint32_t>& b);
+
+private:
+  //! The fastest mod implementation. 
+  uint64_t
+  FastMod(const uint64_t input, const uint64_t ceil) {
```

Comment:
I'm somewhat surprised that this makes a difference at all, even when the only thing a program is doing is `mod`, but it clearly does.
Given the number of other things your code is doing, does this actually have an impact on the runtime of your fingerprinter?

Reply:
It does have an effect, although a minor one.